### ISSUE TITLE
C++ (and other) code review

### ISSUE THREAD

**Issue by LTLA** [jlmelville/uwot#15]
I thought I'd go one by one over each of the C++ files, starting with `nn_parallel.h`. It seems you're doing a nearest-neighbor search on `X` in the R `annoy_nn` function. Here's some observations:

- Annoy will return the observation itself as its own nearest neighbor. I usually have to search `k+1` neighbors and check if the observation is not its own neighbor, see [here](https://github.com/LTLA/BiocNeighbors/blob/master/src/find_annoy.cpp).
- On that note, you can also use the `get_nns_by_item` method for get the NNs for a particular item in the index. This avoids the need to repass `mat` into the `NNworker`, and eliminates the row-by-row accesses that are not cache optimal in `NNworker::operator()`. 
- You may have some small-to-moderate efficiency gains by saving output in columns of `idx` and `dists`, which would be more cache optimal. That is, create transposed matrices for `NNWorker` to dump results in, and untranspose them just before or after you return to R.
- Probably the default grain size should be larger, to avoid false sharing and the like?

**Comment by LTLA** [comment]
Looking at `gradient.cpp` now. It seems like `umap_gradient` and `aumap_gradient` only differ in the power function call. You could avoid code duplication with some templating:

```cpp
// In the header
template<double (*powfun)(double, double)>
class base_umap_gradient {
public:
  umap_gradient(const double a, const double b, const double gamma);
  const double grad_attr(const double dist_squared) const;
  const double grad_rep(const double dist_squared) const;
  // static const constexpr double clip_max = 4.0;
  const double clip_max;
private:
  const double a;
  const double b;
  const double a_b_m2;
  const double gamma_b_2;
};

typedef base_umap_gradient<std::pow> umap_gradient;
double fastPrecisePow(double, double);
typedef base_umap_gradient<fastPrecisePow> aumap_gradient;
```

... and just define methods for `base_umap_gradient` with the templated `powfun`.

Also, if `clip_max` isn't going to be changed, it should be fine to have it as a `constexpr`. If it will be changed, it would be better to do so through a formal setter method.

**Comment by LTLA** [comment]
Continuing with my ~opinions~ facts. On `optimize.cpp`:

- If `rdist` is going to be called many times, you might consider transposing the original matrices `m` and `n` so you can get column-major access. I assume Armadillo stores its matrices in column-major format, but it looks like your distance computations on `m` and `n` are going row-by-row. Doing a one-off transposition at the start can allow column-major access throughout, which is important  for efficient cache access, e.g., I've seen 3-4-fold speed-ups in my _cydar_ package.
- I'm a bit concerned by the seed setting in each `operator()` call. I assume that `parallelFor` controls the number of calls to `operator()` and the identity of the windows processed by each core/worker. If this is the case, the results will not be consistent when a different number of workers is used, or if TBB changes its mind about how the jobs are split across workers. Perhaps define a separate seed for each job right at the start, and then use the appropriate seed in each job loop.

**Comment by jlmelville** [comment]
I really appreciate the attention you've given to this @LTLA. There's a lot for me to process, but will start dealing with this ASAP. I did look at the effect of column-based versus row-based access in the optimization and couldn't detect a difference when I first implemented it. The stochastic nature of the algorithm means we aren't ever iterating over large contiguous chunks of the matrix. But I will revisit this.

Also, I'm not sure that the asynchronous SGD will ever allow for fully repeatable results with more than one thread -- aren't we ultimately at the mercy of the OS as to when different threads are in action? There's no locking of the coordinate matrix and there's no way to cleanly partition the coordinate matrix between threads, so reads and writes happen in an unpredictable order (not to mention the small number of stale reads and lost writes).  I rambled to myself unconvincingly about this at #10.

**Comment by LTLA** [comment]
Hm. This sounds worse than I initially thought.

I assume that the major output of the `SgdWorker::operator()` function is to edit `head_embedding` and `tail_embedding`, but these aren't wrapped in thread-safe classes. I guess this was what you meant by stale reads and lost writes. That seems... pretty bad with respect to algorithmic correctness. 

The more general problem of inconsistent updates will, as you say, result in uncontrollable differences between repeated calls of `uwot`, depending on the data races between threads. It would only be possible to obtain reproducible results when the number of threads is set to 1. This also sounds bad, to the point that I would never use more than one thread in any analysis I planned to present.

The implicit dependencies between jobs makes me wonder whether this function was a good target for parallelization in the first place. There are two obvious paths ahead:

- Take out parallelization for this function. People can still use parallelization for the other tasks, so they don't have to turn it off completely.
- Change the algorithm so that it is read-only in each worker w.r.t. `head_embedding` and `tail_embedding`. Save the results in a separate matrix, as I think you suggested in #10. To me, the extra memory usage is not a big deal compared to getting irreproducible results.

**Comment by LTLA** [comment]
Having looked at the remaining C++ files, it seems like `SmoothKnnWorker` and `PerplexityWorker` have similar parallelization problems due to the updates to `sigma` and `beta` respectively. It might be possible to partition the code into "parts that depend on `sigma`" and everything else, and just parallelize the latter to take the workload off the serial part of the processing.

**Comment by jlmelville** [comment]
As far as I know, the original LargeVis code (where I think this method of asymmetric SGD originates) has exactly the same features. I can't detect any locking from perusing its source, anyway. I'm happy to be corrected on this matter. I don't know anything about Numba, so I can't say how the Python implementation deals with this. Perhaps @lmcinnes knows.

In terms of the badness for algorithmic correctness in term of stale reads and lost writes (by which I mean the classic thread problem of an increment-and-update being non-atomic), it won't happen very often, and grows increasingly rare as the dataset size grows. Gradients are by their nature noisy in SGD, and in practice it doesn't seem like it's an issue. The reproducibility concerns are separate (but valid).

Short term, I am happy to introduce an option to control the number of threads used in the optimization section compared to the rest of the code, if that will be useful (although the speed decrease may make things like FIt-SNE or fastTSNE  a better choice). Fixing this long term... well, it will be a long term project. I can also add a suitable warning to the README to flag this caveat to potential users.

**Comment by jlmelville** [comment]
The SmoothKnn/perplexity work is partitioned per node, so reads and writes to `sigma[i]` and `beta[i]` for a specific `i` only occur in one thread. Or do you mean something else?

**Comment by LTLA** [comment]
Regarding the short term fix: it would probably be better to just stick a "user beware" on any number of threads > 1. My guess is that the run-time is dominated by the NN search (it is for t-SNE, anyway), so even using just 1 thread for everything else would still be pretty quick. A long term solution would require some care, but I think it is worth it at some point. Reproducibility is pretty important.

I'm looking at `smooth_knn.cpp` now, and `sigma` seems to be a `double`.

**Comment by jlmelville** [comment]
> Regarding the short term fix: it would probably be better to just stick a "user beware" on any number of threads > 1. 

Add a warning if more than one thread is specified? Or just log a message to the console if `verbose = TRUE`?

> My guess is that the run-time is dominated by the NN search (it is for t-SNE, anyway), so even using just 1 thread for everything else would still be pretty quick. 

True. Testing on MNIST (although with none of the optimization flags that would be turned on), I estimate about a 15% slow down when going from 6 threads to 1 thread.

> I'm looking at `smooth_knn.cpp` now, and `sigma` seems to be a `double` that gets updated in each loop. So the updates will depend on the identity and number of the loops for each worker.

Oops! That will teach me not to double check. You are correct, sorry. Initializing the sigma search for vertex `i+1` from the final value for `i` value reduces the total number of iterations, but it does indeed mean the converged result will be slightly different depending on the thread boundary. That should be fixed (especially as it's nowhere close to being a bottleneck for usual calculations).

**Comment by jlmelville** [comment]
The annotations on the equivalent Python code do indicate parallelism. I wonder how Numba does it?

https://github.com/lmcinnes/umap/blob/73993282bde371c7e2c9a947c38c586116ad144b/umap/umap_.py#L682

**Comment by LTLA** [comment]
My guess is that numba's parallelization is very low level. Probably occurs only for simple operations that are known to be vectorizable, e.g., calculation of the Euclidean distance.

**Comment by lmcinnes** [comment]
The numba annotation merely tells numba to *try* to parallelise where it can. In practice that means it does an analysis for separable things, which definitely does not include the updates to head and tail embeddings, so that much runs single threaded. This has been among the reasons I haven't made a proper parallel version -- it is hard to get it "correct" in a fully formal way. My impression is that LargeVis does exactly as you say: accept that race conditions will happen and just roll with that. I was never that happy with that approach.

I believe the "right" thing to do would be to do it batch-wise -- have each thread copy the current embedding (or lock it to read only) and perform a batch of update calculations but simply write the proposed updates to a local update buffer. Then aggregate the update buffers from all the threads and do a bulk apply/update to the embedding. This will take more epochs to converge since you are locking the embedding as read only, but would have "correct" multithreading. The major downside with this is that to have consistent results one would have to apply a similar strategy to the single threaded case, and that would definitely negatively impact single threaded performance.

Any thoughts are welcome.

**Comment by jlmelville** [comment]
I have added a new parameter, `n_sgd_threads`, that controls the number of threads used during SGD. By default it is 0, so results are reproducible (in line with the Python UMAP). This, along with the sigma/beta fix, should solve the immediate problem of reproducibility.

I have also made the changes @LTLA  suggested to `gradient.cpp` and it is indeed the case that `mat` did not need to be used with `NNWorker`, and that is also fixed. Using MNIST, I was unable to detect any difference in processing the distance and index matrices in their transposed forms. Nor did the grain size do anything. I will have to start profiling to work out what is going on there.

Fixing `optimize.cpp` may require moving away from both Armadillo and RcppParallel. In the latter case, even with a fixed number of threads, the number of times `operator()` is called is not deterministic. This would mean even by moving to the update-vector implementation, multi-threaded results would not be reproducible due to the RNG.

**Comment by LTLA** [comment]
@jlmelville Good work. Just to motivate the transposition for column-major access:

```r

library(Rcpp)

sourceCpp(code='
#include "Rcpp.h"
#include <algorithm>

// [[Rcpp::export]]
double sum_by_col(Rcpp::NumericMatrix x) {
    double collected=0;
    for (int i=0; i<x.ncol(); ++i) {
        Rcpp::NumericMatrix::Column y=x.column(i);
        collected+=std::accumulate(y.begin(), y.end(), 0.0);
    }
    return collected;
}

// [[Rcpp::export]]
double sum_by_row(Rcpp::NumericMatrix x) {
    double collected=0;
    for (int i=0; i<x.nrow(); ++i) {
        Rcpp::NumericMatrix::Row y=x.row(i);
        collected+=std::accumulate(y.begin(), y.end(), 0.0);
    }
    return collected;
}
')

amat <- matrix(rnorm(4e6), 2e3, 2e3)
library(microbenchmark)
microbenchmark(sum_by_col(amat), sum_by_row(amat))
## Unit: milliseconds
##              expr       min        lq      mean    median        uq      max
##  sum_by_col(amat)  4.092568  4.231072  4.465735  4.364297  4.547842  6.28038
##  sum_by_row(amat) 44.600871 45.531943 46.614799 46.205932 47.197383 54.67396
##  neval
##    100
##    100
```

So processing by column is 10 times faster than processing by row. Pretty neat, huh.

@lmcinnes Your suggested batch-wise procedure makes sense to me - and if the data set was small enough for people to feasibly use a single core, a bit of a slow down probably doesn't matter much.

**Comment by jlmelville** [comment]
These minor changes have also been checked in. Thank you again for such attention to detail @LTLA. At the very least I will have to name the next release after you.

Unfortunately, I have also gone back to passing the items into the `NNWorker`, rather than their ids. The item itself is needed in `umap_transform`, because you are using a search query from the test set using the index built with the training set. I am not sure it's worth maintaining both versions (although they would share a lot of code, so some templating would reduce the duplication).

I will probably look at column-based access in `optimize.cpp` next, although because it never loops over large continuous parts of the matrix, I am not expecting to see the sort of differences in performance the microbenchmark shows.

**Comment by jlmelville** [comment]
About the batched approach to the optimization: I think dealing with the RNG so that the ordering of the stream of random numbers was independent of the number of threads would be pretty difficult.

**Comment by LTLA** [comment]
Regarding the neighbors: probably not worth maintaining both versions, though note the need to search k+1 and the prune the self when you're doing the NN search. Usually this can be done by removing the first neighbor, but sometimes you have ties at zero distance with duplicate points.

Regarding the batching: this should be possible if you seed the RNG separately for each job (not thread). Then, if you define the RNG seeds for all jobs during RNG construction, you'll always get the same set of random numbers within each job regardless of the number of threads they're split over.

**Comment by jlmelville** [comment]
About the k neighbors including the self point, that's intentional: https://github.com/lmcinnes/umap/issues/53 (as you can see, it surprised me too).

**Comment by jlmelville** [comment]
On the batching/RNG issue: I'm still not entirely clear on the RNG strategy and I would very much like to get whatever design help I can on this before enthusiasm for discussing this peters out entirely. I am probably being a bit dense about this, so apologies:

Specifically with RcppParallel I seed the RNG every time `operator(chunk_start, chunk_end)` is called. This is necessary because of the way RcppParallel (and presumably downstream, Intel TBB) works: I can't control the number of times the Worker constructor is called (although for `parallelFor`, that seems to be only once), and it looks like the `operator()` function is called an unpredictable number of times even with a fixed number of threads and `grain_size`. Ultimately I would have to move away from RcppParallel for consistency.

But it's probably safe to assume we will always be partitioning work by splitting up the `epoch_of_next_sample` vector into chunks. Is the suggestion to always have a fixed number of chunks, regardless of the number of threads and have a different RNG with a repeatable but different seed in each chunk of work? Strategies that don't do that presumably have to potentially share an RNG across threads which would lead to locking and/or ordering issues.

**Comment by LTLA** [comment]
Now I'm a bit concerned that I'm not getting the problem... anyway, my idea was:

1. In `SgdWorker::SgdWorker()`, define a seed for each element of `epochs_per_sample`, using the R PRNG or something. This is possible as `n_epochs_per_sample` is known beforehand, and will be trivially consistent across runs as it occurs in a serial part of the code.
2. The seed for each `epochs_per_sample[i]` is used to seed the `tau_prng` PRNG _within_ the `for` loop of `SgdWorker::operator()`. This means that, regardless of how the `epochs_per_sample` jobs are split up, the seeding for each job will not be affected (but still be different from other jobs). This is different to what is currently performed, where the PRNG is seeded outside of the loop.
3. And that's it.

**Comment by LTLA** [comment]
Regarding neighbors: oh. This should probably be noted in the documentation for people who want to provide pre-computed NN results to the function; at least a few of the NN packages I've worked with (and written) will not consider a point to be a neighbor of itself. I've always thought that to be the most intuitive interpretation though I guess there is a mathematical argument for the other approach as well.

**Comment by jlmelville** [comment]
I don't think you misunderstand the optimization RNG issue, I'm just being a bit slow on the up-take. I think you are suggesting that every member of   `epochs_per_sample` has its own seed (or 3 seeds with the current implementation). The negative sampling phase is then effectively done with its own RNG. That would work, as long as the properties of a small number of random numbers generated from many different instantiations of a PRNG work out. Also, this would be repeated at every epoch, so you would need to modify or regenerate the seeds at each epoch. Do I understand correctly?

**Comment by LTLA** [comment]
That sounds right. If the initial seeding in the constructor is done with a high-quality PRNG, the random numbers for each element of `epochs_per_sample` should be okay. I think. It would be simple to just use R's PNRGs for this, with a `Rcpp::RNGScope` object in the constructor. This would allow you to generate new (but consistently reproducible) random numbers at every epoch.

**Comment by LTLA** [comment]
Just to motivate this discussion: https://github.com/davismcc/scater/tree/uwot. So once _uwot_ gets on CRAN, it'll have at least one reverse dependency straight away.

**Comment by LTLA** [comment]
Having actually tried what I suggested... I don't think it works. The seeding scheme works, but the nature of the SGD algorithm causes unresolvable dependencies between workers. 

The closest I got was to make `head_embedding` and `tail_embedding` both read-only; compute steps for every iteration `i` across workers **without modifying the embeddings**; and then somehow "aggregate" the steps across iterations to get a single step per element of the embedding. The problem is with the aggregation when the same `j` occurs in multiple iterations of `i`. Taking the sum is probably inappropriate as a large step for `j` will be repeatedly computed from the read-only embeddings; the correct "total" step will be smaller as steps at earlier `i` should reduce the gradient at later steps when the algorithm can modify the embeddings as it goes along. An additional problem is that we now have to have a matrix of steps for each worker (possibly even each iteration of `i`), which is pretty expensive.

So probably the best solution is to not parallelize the SGD at all. I guess this makes sense in hindsight as this procedure is quite heavily iterative. If `ncol` is large, one might consider parallelizing the inner `d` loops, but that's all I can think of. If you could mini-batch each step of the SGD, that would also provide some opportunity for parallelization, but that is a fairly major refactor.

**Comment by jlmelville** [comment]
I would think that being unable to parallelize the SGD stage of UMAP at all is potentially limiting the method in the long run, so that is disappointing to hear, but I am impressed you implemented and tested this so quickly. 

There might be two solutions to the too-large step problem: a smaller initial learning rate, and to average the updates rather than sum, but the larger structural issues remain.

**Comment by jlmelville** [comment]
With af0f039, the coordinates are stored column-wise for the purposes of optimization. I detected no speed up for the usual case of embedding into two dimensions, but for MNIST and `n_components = 50`, the optimization step was reduced to 11 minutes from 22 minutes.

Also, the dependency on RcppArmadillo has been removed.

**Comment by LTLA** [comment]
I guess we could close this.
